### ISSUE TITLE
Advice on Connection handling

### ISSUE THREAD

**Issue by rristea** [oktal/pistache#95]
Hello,

I am currently using the Pistache client library. And have some use cases that don't seem to be covered out of the box, and I would like some advice on how this can be achieved. I am considering doing some local modifications, but only where necessary.

1) I am starting the Client with a number of concurrent connections to an endpoint (usual initialization). I would like to somehow check if there are any Idle connections before sending a request. If not, then I don't want to send the request.
This is not possible at the moment right?
I did find that this can be achieved by exposing the Client's ConnectionPool, try to do a pool.pickConnection() myself, and if a connection was returned then force a request to use that connection.
But just would like an opinion if this can cause problems.

2) If a Promise is not settled by a certain point I just want to cancel the request. By this I mean reset/close the Connection, and be able to use it for another request.
I did not find this functionality to be possible right now.
What I've tried is to expose the timers somehow, and force timeouts, but failed to get anywhere.
I know that at the moment the timeout functionality is not exposed in the RequestBuilder, but I just implemented locally the setter for the RequestBuilder timeout.
But trying to set timeout to say 100 ms, I get crashes. Even for just one request.
Is the timeout functionality just not exposed yet, or does it also need some more work?
Is there any other way I could safely reset/close a Connection?

Thanks!

**Comment by kennymalac** [comment]
My initial thought on the timeout functionality would be to simply have an interrupt routine that could stop cursor.advance() and would reset the parser state after the time limit has passed

**Comment by rristea** [comment]
Hi. Thanks for the reply.
I need to take a closer look at the parser, but with this solution you assume that the response has already been received, right? and we just stop the parsing.
But it can happen that the Response has not been received yet, and I want to cancel. I'm no longer interested in that response. I would like to close the Connection (i.e. the underlying socket), and open a new one.

**Comment by kennymalac** [comment]
Yeah, that's right. If you want to cancel a request, there'd still be an interrupt, but it needs to take place at an earlier step where the connection is being made. Take a look at https://github.com/oktal/pistache/commit/5730236b80060013f5374c27dd9c662581a5cbaf
Maybe that will help you for submitting a patch to add this functionality?
